Recompute flat_arg_fake_tensors after fakification

andrewor14 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98769

Summary: This fixes the case when some of the input tensors were
real tensors and fakified in `validate_and_convert_non_fake_tensors`,
but `flat_arg_fake_tensors` would not contain all the inputs
because it was computed before the fakification. We fix this by
recomputing `flat_arg_fake_tensors` after fakification as well.

Test Plan:
python test/dynamo/test_export.py ExportTests.test_mixed_real_and_fake_inputs
    
Reviewers: Chillee, voznesenskym

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

andrewor14 committed:
[dynamo] Recompute flat_arg_fake_tensors after fakification

[ghstack-poisoned]

andrewor14 committed:
Update on "[dynamo] Recompute flat_arg_fake_tensors after fakification"




[ghstack-poisoned]

andrewor14 committed:
Update on "Recompute flat_arg_fake_tensors after fakification"


Summary: This fixes the case when some of the input tensors were
real tensors and fakified in `validate_and_convert_non_fake_tensors`,
but `flat_arg_fake_tensors` would not contain all the inputs
because it was computed before the fakification. We fix this by
recomputing `flat_arg_fake_tensors` after fakification as well.

Test Plan:
python test/dynamo/test_export.py ExportTests.test_mixed_real_and_fake_inputs
    
Reviewers: Chillee, voznesenskym

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "Recompute flat_arg_fake_tensors after fakification"


Summary: This fixes the case when some of the input tensors were
real tensors and fakified in `validate_and_convert_non_fake_tensors`,
but `flat_arg_fake_tensors` would not contain all the inputs
because it was computed before the fakification. We fix this by
recomputing `flat_arg_fake_tensors` after fakification as well.

Test Plan:
python test/dynamo/test_export.py ExportTests.test_mixed_real_and_fake_inputs
    
Reviewers: Chillee, voznesenskym

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1161,6 +1161,10 @@ def dispatch(self, func, types, args=(), kwargs=None):
             func, converter, args, kwargs
         )
 
+        # Recompute flat_arg_fake_tensors here again in case some of the inputs
+        # were real tensors and fakified in validate_and_convert_non_fake_tensors
+        flat_arg_fake_tensors = tree_flatten_only(FakeTensor, (args, kwargs))
+
```
This is not free! What is the perf overhead here? Do you need help kicking off a measurement run?

andrewor14 replied:
Yes, that would be great. Thanks.

eellison replied:
You could move the creation of this to `validate_and_convert_non_fake_tensors` ? would be faster ?

eellison reviewed (commented):
lgtm but per @voznesenskym's concerns we should be careful about adding more pytree calls since they are slow

eellison reviewed (commented):
Note that this will prevent real_tensors from being used with constant ops, but I think that's fine

andrewor14 committed:
Update on "Recompute flat_arg_fake_tensors after fakification"


Summary: This fixes the case when some of the input tensors were
real tensors and fakified in `validate_and_convert_non_fake_tensors`,
but `flat_arg_fake_tensors` would not contain all the inputs
because it was computed before the fakification. We fix this by
recomputing `flat_arg_fake_tensors` after fakification as well.

Test Plan:
python test/dynamo/test_export.py ExportTests.test_mixed_real_and_fake_inputs
    
Reviewers: Chillee, voznesenskym

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym commented:
Perf run here https://github.com/pytorch/pytorch/actions/runs/4691840608

andrewor14 commented on test/dynamo/test_export.py:
```diff
@@ -2389,6 +2389,36 @@ def f_correct(x):
+                return conv
+
+        example_inputs = (torch.randn(1, 1, 3, 3),)
+        torch._dynamo.export(
```
Can we do this in a separate PR? This is currently blocking https://github.com/pytorch/pytorch/pull/98568 right now

voznesenskym replied:
Followup is fine by me, I trust you to do it :)

eellison replied:
The fact that there wasn't a good fake tensor only test on the original PR https://github.com/pytorch/pytorch/pull/90417 is probably we have this error lol. Can you please add it in this PR ?

eellison commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1157,7 +1157,7 @@ def dispatch(self, func, types, args=(), kwargs=None):
             return converter(self, args[0])
 
-        args, kwargs = self.validate_and_convert_non_fake_tensors(
+        args, kwargs, flat_arg_fake_tensors = self.validate_and_convert_non_fake_tensors(
```
Can you add back the comment you had earlier ?

andrewor14 replied:
done

eellison reviewed (commented):
looks good but please add separate test in test_fake_tensor

andrewor14 committed:
Update on "Recompute flat_arg_fake_tensors after fakification"


Summary: This fixes the case when some of the input tensors were
real tensors and fakified in `validate_and_convert_non_fake_tensors`,
but `flat_arg_fake_tensors` would not contain all the inputs
because it was computed before the fakification. We fix this by
recomputing `flat_arg_fake_tensors` after fakification as well.

Test Plan:
python test/dynamo/test_export.py ExportTests.test_mixed_real_and_fake_inputs
    
Reviewers: Chillee, voznesenskym

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "Recompute flat_arg_fake_tensors after fakification"


Summary: This fixes the case when some of the input tensors were
real tensors and fakified in `validate_and_convert_non_fake_tensors`,
but `flat_arg_fake_tensors` would not contain all the inputs
because it was computed before the fakification. We fix this by
recomputing `flat_arg_fake_tensors` after fakification as well.

Test Plan:
python test/dynamo/test_export.py ExportTests.test_mixed_real_and_fake_inputs
    
Reviewers: Chillee, voznesenskym

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "Recompute flat_arg_fake_tensors after fakification"


Summary: This fixes the case when some of the input tensors were
real tensors and fakified in `validate_and_convert_non_fake_tensors`,
but `flat_arg_fake_tensors` would not contain all the inputs
because it was computed before the fakification. We fix this by
recomputing `flat_arg_fake_tensors` after fakification as well.

Test Plan:
python test/dynamo/test_export.py ExportTests.test_mixed_real_and_fake_inputs
    
Reviewers: Chillee, voznesenskym

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 committed:
Update on "Recompute flat_arg_fake_tensors after fakification"


Summary: This fixes the case when some of the input tensors were
real tensors and fakified in `validate_and_convert_non_fake_tensors`,
but `flat_arg_fake_tensors` would not contain all the inputs
because it was computed before the fakification. We fix this by
recomputing `flat_arg_fake_tensors` after fakification as well.

Test Plan:
python test/dynamo/test_export.py ExportTests.test_mixed_real_and_fake_inputs
    
Reviewers: Chillee, voznesenskym

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 commented:
Ok, I'm merging this. Thanks everyone for your feedback!

andrewor14 commented:
@pytorchbot merge

andrewor14 commented:
@pytorchbot rebase

andrewor14 committed:
Update on "Recompute flat_arg_fake_tensors after fakification"


Summary: This fixes the case when some of the input tensors were
real tensors and fakified in `validate_and_convert_non_fake_tensors`,
but `flat_arg_fake_tensors` would not contain all the inputs
because it was computed before the fakification. We fix this by
recomputing `flat_arg_fake_tensors` after fakification as well.

Test Plan:
python test/dynamo/test_export.py ExportTests.test_mixed_real_and_fake_inputs
    
Reviewers: Chillee, voznesenskym

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

andrewor14 commented:
@pytorchbot merge
